libevmasm/Assembly.cpp

```diff
@@ -298,6 +505,126 @@ Json::Value Assembly::assemblyJSON(map<string, unsigned> const& _sourceIndices,
+		);
+
+	shared_ptr<Assembly> result = make_shared<Assembly>(langutil::EVMVersion(), _level == 0, "");
+	vector<string> sourceList;
+	if (_json.isMember("sourceList"))
+	{
+		solAssert(_level == 0);
+		for (auto const& it: _json["sourceList"])
+		{
+			solRequire(
+				std::find(sourceList.begin(), sourceList.end(), it.asString()) == sourceList.end(),
+				AssemblyImportException,
+				"Items in 'sourceList' array are not unique."
+			);
+			sourceList.emplace_back(it.asString());
+		}
+	} else
+		sourceList = _sourceList;
+
+	result->importAssemblyItemsFromJSON(_json[".code"], sourceList);
+	if (_json[".auxdata"])
+	{
+		solRequire(_json[".auxdata"].isString(), AssemblyImportException, "Optional member '.auxdata' is not of type string.");
+		bytes auxdata{fromHex(_json[".auxdata"].asString())};
+		solRequire(!auxdata.empty(), AssemblyImportException, "Optional member '.auxdata' is not a valid hexadecimal string.");
+		result->m_auxiliaryData = auxdata;
+	}
+
+	if (_json.isMember(".data"))
+	{
+		solRequire(_json[".data"].isObject(), AssemblyImportException, "Optional member '.data' is not an object.");
+		Json::Value const& data = _json[".data"];
+		for (Json::ValueConstIterator dataIter = data.begin(); dataIter != data.end(); dataIter++)
+		{
+			solRequire(dataIter.key().isString(), AssemblyImportException, "Key inside '.data' is not of type string.");
+			string dataItemID = dataIter.key().asString();
+			Json::Value const& code = data[dataItemID];
+			if (code.isString())
+			{
+				if (!code.asString().empty())
+				{
+					bytes data_value{fromHex(code.asString())};
+					solRequire(
+						!data_value.empty(),
+						AssemblyImportException,
+						"Member '.data' contains a value for '" + dataItemID + "' that is not a valid hexadecimal string.");
+				}
+				result->m_data[h256(fromHex(dataItemID))] = fromHex(code.asString());
+			}
+			else if (code.isObject())
+			{
+				shared_ptr<Assembly> subassembly(Assembly::fromJSON(code, sourceList, _level + 1).first);
+				solAssert(subassembly);
+				result->m_subs.emplace_back(subassembly);
+			}
+			else
+				solThrow(AssemblyImportException, "Key inside '.data' '" + dataItemID + "' can only be a valid hex-string or an object.");
+		}
+	}
+	if (_level == 0)
+		result->updatePaths();
+	return std::make_pair(result, sourceList);
+}
+
+void Assembly::updatePaths(std::vector<Assembly*> const& _parents, std::vector<size_t> const& _absolutePathFromRoot)
+{
+	size_t subId = 0;
+	for (auto& assembly: this->m_subs)
+	{
+		std::vector<Assembly*> parents{_parents};
+		parents.push_back(this);
+
+		std::vector<size_t> absolutePathFromRoot{_absolutePathFromRoot};
+		absolutePathFromRoot.emplace_back(subId);
+
+		int pindex = 0;
+		for (auto& parent: parents)
+		{
+			if (pindex == 0)
+				parent->encodeSubPath(absolutePathFromRoot);
+			else
+			{
+				std::vector<size_t> relativePath{absolutePathFromRoot};
+				for (int i = 0; i < pindex; ++i)
+					relativePath.erase(relativePath.begin());
```

Comment:
```suggestion
				vector<size_t> relativePath{absolutePathFromRoot.begin() + pindex, absolutePathFromRoot.end()};
```
I think this produces the same result.